flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java

```diff
@@ -74,22 +79,46 @@
 			@Nonnull ResourceManager<?> resourceManager,
 			@Nonnull LeaderRetrievalService dispatcherLeaderRetrievalService,
 			@Nonnull LeaderRetrievalService resourceManagerRetrievalService,
-			@Nonnull WebMonitorEndpoint<?> webMonitorEndpoint) {
+			@Nonnull WebMonitorEndpoint<?> webMonitorEndpoint,
+			@Nonnull FatalErrorHandler fatalErrorHandler,
+			@Nonnull CompletableFuture<DispatcherGateway> dispatcherGatewayCompletableFuture) {
 		this.dispatcherRunner = dispatcherRunner;
 		this.resourceManager = resourceManager;
 		this.dispatcherLeaderRetrievalService = dispatcherLeaderRetrievalService;
 		this.resourceManagerRetrievalService = resourceManagerRetrievalService;
 		this.webMonitorEndpoint = webMonitorEndpoint;
+		this.fatalErrorHandler = fatalErrorHandler;
 		this.terminationFuture = new CompletableFuture<>();
 		this.shutDownFuture = new CompletableFuture<>();
 
 		registerShutDownFuture();
+		failOnPrematureTermination(dispatcherGatewayCompletableFuture);
 	}
 
 	private void registerShutDownFuture() {
 		FutureUtils.forward(dispatcherRunner.getShutDownFuture(), shutDownFuture);
 	}
 
+	private void failOnPrematureTermination(CompletableFuture<DispatcherGateway> dispatcherGatewayCompletableFuture) {
+		dispatcherGatewayCompletableFuture.whenComplete((dispatcher, throwable) -> {
+			if (dispatcher != null && dispatcher instanceof Dispatcher) {
```

Comment:
This breaks the DispatcherGateway abstraction and reinforces the assumption that the RM and Dispatcher run in the same JVM which may change in the future.
Why can we not use the shutdown future of the `DispatcherRunner`?

Reply:
The shutdown future of the `DispatcherRunner` has been registed with the shutDown Future of `DispatcherResourceManagerComponent`.  I think there should be no need to deal with it again.
How about that only deals with the Termination Future of ResourceManager here?

Reply:
Are the DispatcherRunner and Dispatcher termination futures linked in any way? As in, if the Dispatcher shuts down _at any point_, then it is guaranteed that the DispatcherRunner termination future is also being completed?

Reply:
if we use the shutdown future of the DispatcherRunner like this
```
private void failOnPrematureTermination() {
	CompletableFuture.anyOf(dispatcherRunner.getShutDownFuture(), resourceManager.getTerminationFuture())
		.whenComplete((ignored, t) -> {
			if (t == null) {
				LOG.error("Dispatcher/ResourceManager shut down because something unexpected happen!");
			}
			if (isRunning.get()) {
				LOG.warn("DispatcherResourceManagerComponent shut down because the Dispatcher/ResourceManager unexpected terminated.");
				fatalErrorHandler.onFatalError(t);
			}
		});
}
```
in per-job mode, `ClusterEntrypoint.shutDownAsync()` may not have a chance to execute, because `ClusterEntrypoint.onFatalError` will call `System.exit(RUNTIME_FAILURE_RETURN_CODE);`
Do you have any good suggestions for revision?

Reply:
We first execute `failOnPrematureTermination();` and then execute `registerShutDownFuture();` 
How about this?  @zentol